Fix support for local functions in "Make Method Asynchronous"

sharwell opened the pull request:
Fixes #14133 

I also fixed a similar issue in the "Make Method Synchronous" code fix provider.

Tagging @dotnet/roslyn-ide  for review

dpoeschl commented on src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs:
```diff
@@ -599,5 +599,88 @@ class Program
+        return 1;
+    }
+}",
+index: 1);
```
This is super subtle. So subtle I don't get it. All of the code in these two tests is the same right? Are there 2 code actions here that result in the same thing? What am I missing?

dpoeschl replied:
I think the non-string arguments are usually indented to the normal level. Example: http://source.roslyn.io/#Roslyn.Services.Editor.CSharp.UnitTests/AddBraces/AddBracesTests.cs,227

CyrusNajmabadi replied:
We're inconsistent with that.  I'm fine with this indentation.

sharwell replied:
@dpoeschl One of them changes the return type to `Task`, while the other leaves it as `void`.

dpoeschl replied:
Can you name the test method names to match that description?

sharwell replied:
💭  Originally they were but then I changed them to match the less descriptive naming indicated by @CyrusNajmabadi in a different PR.

dpoeschl replied:
I'd switch back to your original names, especially given how subtle the difference is between the two tests. There are plenty of examples of descriptive test names, way more than `TestWhatever#` examples. I'm sure @CyrusNajmabadi will agree. :smile:

sharwell replied:
➡️ Renamed the tests.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs:
```diff
@@ -599,5 +599,88 @@ class Program
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsMakeMethodAsynchronous)]
+        [WorkItem(14133, "https://github.com/dotnet/roslyn/issues/14133")]
+        public async Task AddAsyncInLocalFunction2()
```
i don't see any difference between this and the previous test.

CyrusNajmabadi replied:
Can we have a test with comment trivia before the local function.

sharwell replied:
:arrow_right: Added 18 tests. Fixed behavior for some and left two skipped per #18396.

CyrusNajmabadi commented on src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs:
```diff
@@ -19,14 +19,17 @@ internal class CSharpMakeMethodSynchronousCodeFixProvider : AbstractMakeMethodSy
 
         protected override bool IsMethodOrAnonymousFunction(SyntaxNode node)
```
rename method to match current behavior.

sharwell replied:
:arrow_right: See previous case.

CyrusNajmabadi commented on src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs:
```diff
@@ -35,7 +35,9 @@ protected override string GetMakeAsyncVoidFunctionResource()
 
         protected override bool IsMethodOrAnonymousFunction(SyntaxNode node)
```
rename method to match behavior.

sharwell replied:
:arrow_right: Renamed to `IsContainingFunction`

CyrusNajmabadi commented on src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs:
```diff
@@ -35,7 +35,9 @@ protected override string GetMakeAsyncVoidFunctionResource()
-            return node.IsKind(SyntaxKind.MethodDeclaration) || node.IsAnyLambdaOrAnonymousMethod();
+            return node.IsKind(SyntaxKind.MethodDeclaration)
+                || node.IsAnyLambdaOrAnonymousMethod()
+                || node.IsKind(SyntaxKind.LocalFunctionStatement);
```
duplicated code.  can we move to an extension?  (i actually think we have one that already does this somewhere).

sharwell replied:
I couldn't find an existing extension for this. I always wasn't sure what to call it or where to put it if we did have it. For now I left it but if you have a more specific suggestion I can implement that. The functionality does seem specific to just these two operations, at least for now.

CyrusNajmabadi commented on src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs:
```diff
@@ -51,7 +51,7 @@ internal abstract class AbstractMakeMethodAsynchronousCodeFixProvider : CodeFixP
             // If it's a void returning method, offer to keep the void return type, or convert to 
             // a Task return type.
-            if (symbol?.MethodKind == MethodKind.Ordinary &&
+            if ((symbol?.MethodKind == MethodKind.Ordinary || symbol?.MethodKind == MethodKind.LocalFunction) &&
```
please don't mix ||s and &&s in expressions.  Just use nested ifs.  thanks.

CyrusNajmabadi replied:
alternatively, extract method to do the || check, or introduce local.

sharwell replied:
:arrow_right: Introduced local.

CyrusNajmabadi commented on src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs:
```diff
@@ -100,7 +100,7 @@ internal abstract class AbstractMakeMethodAsynchronousCodeFixProvider : CodeFixP
             var methodSymbolOpt = semanticModel.GetDeclaredSymbol(node) as IMethodSymbol;
 
-            if (methodSymbolOpt?.MethodKind == MethodKind.Ordinary &&
+            if ((methodSymbolOpt?.MethodKind == MethodKind.Ordinary || methodSymbolOpt?.MethodKind == MethodKind.LocalFunction) &&
```
same.

sharwell replied:
:arrow_right: Introduced local.

CyrusNajmabadi commented on src/Features/Core/Portable/MakeMethodSynchronous/AbstractMakeMethodSynchronousCodeFixProvider.cs:
```diff
@@ -49,7 +49,7 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
             var methodSymbolOpt = semanticModel.GetDeclaredSymbol(node) as IMethodSymbol;
 
-            if (methodSymbolOpt?.MethodKind == MethodKind.Ordinary &&
+            if ((methodSymbolOpt?.MethodKind == MethodKind.Ordinary || methodSymbolOpt?.MethodKind == MethodKind.LocalFunction) &&
```
same.

sharwell replied:
:arrow_right: Introduced local.

sharwell commented:
@CyrusNajmabadi Can you check the updated handling of trivia (9857a18, 6e6001b) and the renamed methods (d8eac62)?

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Diagnostics/MakeMethodAsynchronous/MakeMethodAsynchronousTests.cs:
```diff
@@ -599,5 +599,180 @@ class Program
+index: 1);
+        }
+
+        [Theory]
```
i'm really not liking this Theory approach to these sorts of tests.  Reading the tests is now much harder.  And it makes basic stuff like { } in the test become distracting as {{ and }}   I don't see why this is actually better than just having two actual tests here.  One for the Task case and one for the void case.

A real purpose for these tests is for the text of the test to be self contained and self describing.  i.e. a single string that contains all the information necessary to know what's going on.  That's why we have the markup language.  There is no separate mechanism for specifying the caret position.  It's just in the text.  This 'Theory' approach fights against that, with negative results IMO.

I think theories can be good for very basic, non-replacing tests.  Where the theory/data contain all the information you need to know.  In that sense, it's a way of stamping out tests easily, but each test is still clear on what it does.  As that's not the case here 👎 on this part for me.

sharwell replied:
I find the theory offers two benefits:

1. It's a form of duplicate code elimination, not unlike the other cases you requested
2. It helps ensure that you can't accidentally introduce slightly different behavior in two closely related cases. For example, this test revealed that changing a token from `Task` to `void` altered the behavior of the formatting engine somewhere. It's much harder to see these differences when the tests are separated.

Considering you get used to the `{{` and `}}` forms pretty quickly (it's no worse than `[|text|]` plus the code samples are small) and the editor highlights the replacements, I find this test form both clearer and more robust for handling a coupled test case like this.

@Pilchie what do you think?

Pilchie replied:
mailto:lrieger@precast-software.com

Pilchie replied:
I'm tentatively with @sharwell on this one, though I also find the `{{` to be tough to read, particularly in a GitHub diff.

Not sure if non-interpolated strings with single `{` and concats with ` + ` would be better though.

CyrusNajmabadi replied:
1. It's a form of duplicate code elimination

I don't see that as a virtue here.  For many test suites I could imagine parameterizing everything to the point that all you do is call a test method with something like "Test("public", "Task", "async", inMethod: true, isLocal: true, useSystem: true)" etc. etc.  Thse would just stamp out all the different variations and test the codepaths.  However, to me, the understandability of the test is gone.  I'm very much in favor of the test acting *completely* as the expected input and expected output.  In a way it acts like a spec would, saying: given a user providing precisely this, I expect to get precisely that.

Yes, lot so of the tests will have overlapping sections.  But it does not feel virtuous to be extracting that out.

CyrusNajmabadi commented on src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs:
```diff
@@ -17,16 +17,19 @@ internal class CSharpMakeMethodSynchronousCodeFixProvider : AbstractMakeMethodSy
-            return node.IsKind(SyntaxKind.MethodDeclaration) || node.IsAnyLambdaOrAnonymousMethod();
+            return node.IsKind(SyntaxKind.MethodDeclaration)
+                || node.IsAnyLambdaOrAnonymousMethod()
+                || node.IsKind(SyntaxKind.LocalFunctionStatement);
```
duplicate code.  we really should have an extension for this.

sharwell replied:
Implemented as `IsAsyncSupportingFunctionSyntax`

CyrusNajmabadi commented on src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs:
```diff
@@ -36,47 +39,66 @@ protected override SyntaxNode RemoveAsyncTokenAndFixReturnType(IMethodSymbol met
         {
-            var newReturnType = method.ReturnType;
+            var newReturnType = FixMethodReturnType(methodSymbol, method.ReturnType, taskType, taskOfTType);
+            SyntaxTokenList newModifiers = FixMethodModifiers(method.Modifiers, ref newReturnType);
```
var

CyrusNajmabadi commented on src/Features/CSharp/Portable/MakeMethodSynchronous/CSharpMakeMethodSynchronousCodeFixProvider.cs:
```diff
@@ -36,47 +39,66 @@ protected override SyntaxNode RemoveAsyncTokenAndFixReturnType(IMethodSymbol met
+        private SyntaxNode FixLocalFunction(IMethodSymbol methodSymbol, LocalFunctionStatementSyntax localFunction, ITypeSymbol taskType, ITypeSymbol taskOfTType)
+        {
+            var newReturnType = FixMethodReturnType(methodSymbol, localFunction.ReturnType, taskType, taskOfTType);
+            SyntaxTokenList newModifiers = FixMethodModifiers(localFunction.Modifiers, ref newReturnType);
```
var

CyrusNajmabadi replied:
how about taking in the return type, and returning it as a tuple with the modifiers.  Not a huge fan of 'ref' unless necessary for perf.

sharwell replied:
:memo: I didn't change the `ref` usage in this case. This usage was a side effect of the simple fact that `SyntaxTokenList` is a value type.

CyrusNajmabadi commented on src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs:
```diff
@@ -100,8 +101,10 @@ internal abstract class AbstractMakeMethodAsynchronousCodeFixProvider : CodeFixP
-                !methodSymbolOpt.Name.EndsWith(AsyncSuffix))
+            bool isOrdinaryOrLocalFunction =
+                methodSymbolOpt?.MethodKind == MethodKind.Ordinary
+                || methodSymbolOpt?.MethodKind == MethodKind.LocalFunction;
```
our general pattern is to keep && and || with the left side of the binary and to align the left and right when they wrap.

CyrusNajmabadi commented on src/Features/Core/Portable/MakeMethodAsynchronous/AbstractMakeMethodAsynchronousCodeFixProvider.cs:
```diff
@@ -51,8 +50,10 @@ internal abstract class AbstractMakeMethodAsynchronousCodeFixProvider : CodeFixP
-                symbol.ReturnsVoid)
+            bool isOrdinaryOrLocalFunction =
+                symbol?.MethodKind == MethodKind.Ordinary
+                || symbol?.MethodKind == MethodKind.LocalFunction;
```
triplicated code.  please extract helper.

sharwell replied:
➡️ Implemented as `IsOrdinaryMethodOrLocalFunction`

sharwell commented:
@CyrusNajmabadi I rebased this PR, and corrected the use of `var` inside one of the existing commits.

The actually new commits relevant to fixing your comments are:

* 09f3547178070a55a98c8d2187799c5c799781b9: Extract helper `IsOrdinaryMethodOrLocalFunction`
* 1bb2cbd2fb38efbfd6df5764b8d417f6e4836f12: Extract helper `IsAsyncSupportingFunctionSyntax`

sharwell commented:
@MattGertz for ask mode

## Ask Mode

**Customer scenario**

Use the Make Method Asynchronous or Make Method Synchronous refactoring operations inside of a local function.

**Bugs this fixes:**

#14133

**Workarounds, if any**

Perform the operation manually.

**Risk**

Low; code changes are restricted to this refactoring operation. Existing tests were unaltered and new tests for several cases were added.

**Performance impact**

Negligible. The algorithmic complexity did not change, and no additional allocations on the fast path (user takes no action), or on the existing path (action does not involve local functions).

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

Failed to add tests for a new language feature to an existing refactoring operation.

**How was the bug found?**

Internal user reported.

sharwell commented:
@dotnet-bot retest this please

MattGertz commented:
Approved pending test completions.
